python/ml4ir/applications/ranking/tests/test_ml4ir_sanity.py

```diff
@@ -0,0 +1,175 @@
+    for clickedDocIdx in clickedDocIdxs:
+        clickedDocScore = y_pred[clickedDocIdx]
+        # ties are resolved with the worst case (i.e. the correct doc will have the lowest place among other docs with same score)
+        equalScoreCount = sortedPred.count(clickedDocScore)
+        predictedRank = sortedPred.index(clickedDocScore) + equalScoreCount - 1.0
+        mrr += 1.0 / (predictedRank + 1.0)
+    mrr = mrr / len(clickedDocIdxs)
+    return mrr
+
+
+def predict(model, documents):
+    """
+    Given a linear model and a matrix of documents, returns the prediction score for each document.
+    """
+    return np.dot(documents, model)
+
+
+def check_mrr(df, model):
+    """
+    calculated MRR for a dataset given a model.
+    """
+    featureCount = len(model)
+    MRR = 0
+    queries = df.groupby('query_id')
+    for qid, group in queries:
+        X = [group['f' + str(i)] for i in range(featureCount)]
+        pred = predict(model, np.array(X).T)
+        mrr = calculate_mrr(pred, group['clicked'])
+        MRR += mrr
+    return MRR / len(queries)
+
+
+def ml4ir_sanity_pipeline(df, perceptron, log_reg, working_dir, log_dir, n_features):
+    """
+    Train ml4ir on the passed data and calculate the MRR for ml4ir, perceptron and logistic regression models.
+    """
+    df.to_csv(working_dir / 'train' / 'data.csv')
+    df.to_csv(working_dir / 'validation' / 'data.csv')
+    df.to_csv(working_dir / 'test' / 'data.csv')
+
+    fconfig_name = "feature_config_sanity_tests_" + str(n_features) + "_features.yaml"
+    feature_config_file = pathlib.Path(__file__).parent / "data" / "configs" / fconfig_name
+    model_config_file = pathlib.Path(__file__).parent / "data" / "configs" / "model_config_sanity_tests.yaml"
+    train_ml4ir(working_dir.as_posix(), feature_config_file.as_posix(), model_config_file.as_posix(), log_dir.as_posix())
+
+    ml4ir_weights = pd.read_csv("models/test_command_line/coefficients.csv")["weight"].tolist()
+    ml4ir_weights = np.array(ml4ir_weights)
+    ml4ir_mrr = check_mrr(df, ml4ir_weights)
+    perceptron_mrr = check_mrr(df, perceptron)
+    log_regression_mrr = check_mrr(df, log_reg)
+    return ml4ir_mrr, perceptron_mrr, log_regression_mrr
+
+
+def train_ml4ir(data_dir, feature_config, model_config, logs_dir):
+    """
+    Train a pointwise ranker, listwise loss model using ml4ir
+    """
+    argv = ['--data_dir', data_dir, '--feature_config', feature_config, '--loss_type', "listwise", '--scoring_type',
+            "listwise",
+            '--run_id', 'test_command_line', '--data_format', 'csv', '--execution_mode', 'train_inference_evaluate',
+            '--loss_key', 'rank_one_listnet',
+            '--num_epochs', '150', '--model_config', model_config, '--batch_size', '1', '--logs_dir', logs_dir,
+            '--max_sequence_size', '25', '--train_pcent_split', '1.0', '--val_pcent_split', '-1', '--test_pcent_split',
+            '-1', '--early_stopping_patience', '1000']
```

Comment:
@mohazahran You should indent these things. Otherwise it is a nightmare to read and to play with.
```suggestion
    argv = ["--data_dir", data_dir,
            "--feature_config", feature_config,
            "--loss_type", "listwise",
            "--scoring_type", "listwise",
            "--run_id", "test_command_line",
            "--data_format", "csv",
            "--execution_mode", "train_inference_evaluate",
            "--loss_key", "rank_one_listnet",
            "--num_epochs", "150",
            "--model_config", model_config,
            "--batch_size", "1",
            "--logs_dir", logs_dir,
            "--max_sequence_size", "25",
            "--train_pcent_split", "1.0",
            "--val_pcent_split", "-1",
            "--test_pcent_split", "-1",
            "--early_stopping_patience", "25",
            "--metrics_keys", "MRR", "categorical_accuracy",
            "--monitor_metric", "categorical_accuracy"]
```

You will notice that I have used one trick here - I have set the `monitor_metric` to `categorical_accuracy`. This takes the MRR and accuracy closer to logistic regression for listwise loss. 

And again, I think your success criteria is too strict, especially given the fact that there are only about 40 queries. I would use `np.isclose(..., atol=0.01)`